cryptosolver.py

```diff
@@ -13,39 +13,97 @@
 import sys
 import time
 
+def parse_constraint_input(constraint_string):
+    # give a constraint string of comma-separated KeyValue characters: "AB,CD,EF"
+    # generate a partial solution dictionary: { A:B, C:D, E:F }
+    solution = {}
+    pairs = constraint_string.upper().split(",")
+    known_keys = []
+    for pair in pairs:
+        if len(pair) == 2:
+            key = pair[0]
+            value = pair[1]
+            if key not in known_keys:
+                solution[key] = value
+                known_keys.append(key)
+    return solution
+
+def parse_known_values(constraint_string):
+    # give a constraint string of comma-separated KeyValue characters: "AB,CD,EF"
+    # generate a list of known values: [ B, D, F ]
+    pairs = constraint_string.upper().split(",")
+    known_values = []
+    for pair in pairs:
+        if len(pair) == 2:
+            key = pair[0]
+            value = pair[1]
+            if value not in known_values:
+                known_values.append(value)
+    return known_values
+
 def compute_fingerprint(word):
     # Given a 'word': compute a canonical "fingerprint" of the characters within
-    # where each character is assigned a letter of the alphabet representing its
-    # sequence of "first occurrence".
+    # where each unknown character is assigned a lowercase letter of the alphabet 
+    # representing its sequence of "first occurrence", and
+    # each known character is assigned its value in uppercase.
     #
     # For example: consider the word 'FALLOW'
     #
-    # 'F' shows up first and is assigned 'A'
-    # 'A' shows up second and is assigned 'B'
-    # 'L' shows up third and is assigned 'C'
+    # 'F' shows up first and is assigned 'a'
+    # 'A' shows up second and is assigned 'b'
+    # 'L' shows up third and is assigned 'c'
     # ... and so on D, E, F...
     #
-    # Its fingerprint would be 'ABCCDE'
+    # Its fingerprint would be 'abccde'
     #
-    # The word 'FELLOW' would have the same fingerprint 'ABCCDE'
-    # however 'FOLLOW' would be different: 'ABCCBD'
+    # The word 'FELLOW' would have the same fingerprint 'abccde'
+    # however 'FOLLOW' would be different: 'abccbd'
     #
-    assignments = {}
+    # For a scrambled word BPKKYU if we knew that K:A,
+    # its fingerprint would be 'abLLcd'
+    #
+    assignments = constraint.copy()
     fp = ""
     char_index = 0
     for char in word.upper():
         if (char >= 'A' and char <= 'Z'):
             if char in assignments:
                 fp += assignments[char]
             else:
-                letter = chr(ord('A') + char_index)
+                letter = chr(ord('a') + char_index)
                 assignments[char] = letter
                 fp += letter
                 char_index += 1
         else:
             fp += char
     return fp
```

Comment:
Yes this `compute_fingerprint()` above and `compute_candidate_fingerprint()` below look good.  Each uses a variable from the global scope (`constraint` in the former and `known_values` in the latter) which is something I would normally avoid (e.g. I would prefer to pass them explicitly as arguments) for reasons of: clearer more readable code, improved long-term maintainability of code.  But not a big deal for a small project like this.